wsmux/stream.go

```diff
@@ -0,0 +1,237 @@
+package wsmux
+
+import (
+	"io"
+	"net"
+	"sync"
+	"time"
+)
+
+const (
+	DefaultCapacity = 1024
+)
+
+type stream struct {
+	id        uint32
+	m         sync.Mutex
+	c         *sync.Cond
+	b         *buffer
+	unblocked uint32
+
+	endErr error
+
+	closed       bool
+	remoteClosed bool
+	accepted     chan struct{}
+
+	session *Session
+
+	//deadlines
+	writeDeadline time.Time
+	readDeadline  time.Time
+}
+
+func newStream(id uint32, session *Session) *stream {
+	str := &stream{
+		id:        id,
+		b:         newBuffer(DefaultCapacity),
+		unblocked: 0,
+
+		endErr: nil,
+
+		closed:       false,
+		remoteClosed: false,
+		accepted:     make(chan struct{}),
+
+		session: session,
+	}
+
+	str.c = sync.NewCond(&str.m)
```

Comment:
You can also do:
`str.c.L = &str.m`

If you use `sync.Cond` instead of `*sync.Cond`... That means one less pointer to play with...